turbo/consumers.py

```diff
@@ -42,21 +39,32 @@ def notify(self, event, *args, **kwargs):
                     model_name.lower(): instance,
                     "action": action,
                     "dom_target": dom_target,
-                    "model_template": f"{app}/{model_name}.html"
+                    "model_template": template,
+                    **context,
                 })
             })
 
     def receive_json(self, content, **kwargs):
-        model_label = content.get("model")
-        pk = content.get("pk")
+        signer = Signer()
         request_id = content.get("request_id")
+        if request_id is None:
+            raise TurboStreamException("No request_id in subscription request.")
+        message_type = content.get("type")
+        if message_type == "subscribe":
+            try:
+                channel_name = signer.unsign(content.get("signed_channel_name", ""))
+            except BadSignature:
+                raise TurboStreamException("Signature has been tampered with on the client!")
 
-        channel_name = make_channel_name(model_label, pk)
+            self.requests.setdefault(channel_name, []).append(request_id)
+            self.groups.append(channel_name)
+            async_to_sync(self.channel_layer.group_add)(channel_name, self.channel_name)
+        elif message_type == "unsubscribe":
+            try:
+                channel_name = [channel_name for channel_name, requests in self.requests.items() if request_id in requests][0]
+            except IndexError:
+                raise TurboStreamException("No subscription for a given request ID exists to unsubscribe.")
```

Comment:
Do we want to raise an error here ? Not finding the subscription to unsubscribe does not seem like a big deal to me (after all, the client is already unsubscribed if we can't find their subscription , right ?). Maybe just logging an "info-level" message would be enough ?

Reply:
This should be an unreachable state unless the client sends the subscribe and unsubscribe messages out-of-order, so I do think this makes sense as either an error or a log warning, although @blopker if you want to jump in here too since the error was your suggestion?

Reply:
I was on the fence about this as well, but ultimately I'd want to know if some supposedly unreachable code gets run. Best case it never happens, worst case our assumptions are currently wrong or something changes in the future to make them wrong. I'd want to know if and when that happens. In this case we'll learn something and the client wont really be affected.